src/frontend/src/handler/query.rs

```diff
@@ -12,61 +12,103 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::sync::Arc;
+
 use futures::StreamExt;
 use pgwire::pg_field_descriptor::PgFieldDescriptor;
 use pgwire::pg_response::{PgResponse, StatementType};
 use risingwave_common::error::{ErrorCode, Result, RwError};
 use risingwave_common::session_config::QueryMode;
 use risingwave_sqlparser::ast::Statement;
-use tracing::debug;
 
 use super::{PgResponseStream, RwPgResponse};
-use crate::binder::{Binder, BoundStatement};
+use crate::binder::Binder;
 use crate::handler::privilege::{check_privileges, resolve_privileges};
 use crate::handler::util::{force_local_mode, to_pg_field};
 use crate::planner::Planner;
+use crate::scheduler::plan_fragmenter::Query;
 use crate::scheduler::{
     BatchPlanFragmenter, ExecutionContext, ExecutionContextRef, LocalQueryExecution,
 };
-use crate::session::{OptimizerContext, SessionImpl};
+use crate::session::{OptimizerContext, OptimizerContextRef, SessionImpl};
+use crate::PlanRef;
 
 pub type QueryResultSet = PgResponseStream;
 
-pub async fn handle_query(
-    context: OptimizerContext,
+pub fn gen_batch_query_plan(
+    session: &SessionImpl,
+    context: OptimizerContextRef,
     stmt: Statement,
-    format: bool,
-) -> Result<RwPgResponse> {
+) -> Result<(PlanRef, QueryMode, Vec<PgFieldDescriptor>)> {
     let stmt_type = to_statement_type(&stmt);
-    let session = context.session_ctx.clone();
 
     let bound = {
-        let mut binder = Binder::new(&session);
+        let mut binder = Binder::new(session);
         binder.bind(stmt)?
     };
 
     let check_items = resolve_privileges(&bound);
-    check_privileges(&session, &check_items)?;
+    check_privileges(session, &check_items)?;
 
-    let query_mode = if force_local_mode(&bound) {
-        debug!("force query mode to local");
-        QueryMode::Local
-    } else {
-        session.config().get_query_mode()
-    };
-    debug!("query_mode:{:?}", query_mode);
-
-    let (mut row_stream, pg_descs) = match query_mode {
-        QueryMode::Local => {
-            if stmt_type.is_dml() {
-                // DML do not support local mode yet.
-                distribute_execute(context, bound, format).await?
-            } else {
-                local_execute(context, bound, format).await?
-            }
+    let mut planner = Planner::new(context);
+
+    let must_local = force_local_mode(&bound);
+    let must_dist = stmt_type.is_dml();
```

Comment:
Why not put this logic in `force_local_mode`?

Reply:
I will remove the `force_local_mode` because now this is the only place using it :smile_cat: